### PR TITLE
[DiagnosticsSema] SR-5983

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Improve unused result warning for closures.

Resolves [SR-5983](https://bugs.swift.org/browse/SR-5983).

### CODE REVIEW

**Comment by CodaFi** [comment]
You’ll need to update the tests as well.

**Comment by AnthonyLatsis** [comment]
Hi Robert,

Could you please elaborate? This is my first PR. 

Thank you in advance.

**Comment by CodaFi** [comment]
It’s not enough to just change the text of a diagnostic, we have a number of tests that are testing for the old diagnostic to be fired.  You will need to search for these tests and update their expectations to your new wording.

You can invoke the tests by either running a full rebuild with the `test` flag

> $ ./utils/build-script --test

Or by manually invoking `lit` in the appropriate part of the Swift tests folder in your build directory.

**Comment by AnthonyLatsis** [comment]
Found 5 matches, they were all commented. I made sure there aren't any typos, but I haven't run the tests. I searched for `lit` and found three `lit.site.cfg` files, tried running all of them but ended up with a compile-time error. Either the python compiler version is inappropriate (2.7.10), or I'm doing something wrong.

**Comment by CodaFi** [comment]
> Found 5 matches, they were all commented

The verifier (which you can run by adding the `-verify` flag to a `swift` invocation) is comment-based in the tradition of other LLVM tools like `lit` and `FileCheck`.  `expected-warning` and `expected-error` are assertions that a given diagnostic is fired on that line given the code that is there as well.

> I searched for lit

`lit` exists in LLVM’s tree, not Swift’s.  Assuming you are focused on the directory of your Swift checkout, and you built with Ninja in debug mode, this sample command should work 

```
$ ../llvm/utils/lit/lit.py ../build/Ninja-DebugAssert/swift-macosx-x86_64/test-macosx-x86_64
```

**Comment by AnthonyLatsis** [comment]
Ah, so you meant invoking `lit` for a certain directory.

> The verifier (which you can run by adding the -verify flag to a swift invocation) is comment-based in the tradition of other LLVM tools like lit and FileCheck. expected-warning and expected-error are assertions that a given diagnostic is fired on that line given the code that is there as well.

Does this mean commented assertions are actually processed?

I will invoke the tests to verify everything is alright. Is that enough?

**Comment by AnthonyLatsis** [comment]
> not necessarily correct

I assume you meant that it would be nice to be more concrete (a closure is a function after all).

Would adding a `%select{function|closure}` to the diagnostic and handling it be a valid approach?

**Comment by CodaFi** [comment]
Exactly. Whichever you choose to implement is fine.

**Comment by AnthonyLatsis** [comment]
A quick question: when compiling after minor changes, linking always takes the most time (~> 5 minutes). Is this the way it should be or is there a way to speed it up?

**Comment by CodaFi** [comment]
Are you running an incremental build or are you invoking the build script every time?

**Comment by AnthonyLatsis** [comment]
> /Volumes/Elements/swift-source/llvm/utils/lit

Incremental of course. I work with xcode

File: test/Parse/recovery.swift [hunk 1]
```diff
@@ -1,783 +0,0 @@
-// RUN: %target-typecheck-verify-swift
```
Comment (CodaFi) [inline test/Parse/recovery.swift]: Please restore this test

**Comment by CodaFi** [comment]
Linking swift takes around 30 seconds for me.  I suspect it's a function of hardware if the full build is also taking 5 hours.

**Comment by AnthonyLatsis** [comment]
I have no idea how that happened, my apologies.

File: lib/Sema/TypeCheckStmt.cpp [hunk 2]
```diff
@@ -1322,9 +1322,9 @@ void TypeChecker::checkIgnoredExpr(Expr *E) {
         .highlight(SR1).highlight(SR2);
     } else
       diagnose(fn->getLoc(), diag::expression_unused_result_unknown,
-               valueE->getType())
+               isa<ClosureExpr>(fn), valueE->getType())
         .highlight(SR1).highlight(SR2);
-
+
```
Comment (CodaFi) [inline lib/Sema/TypeCheckStmt.cpp]: Last nit: Eliminate this little bit of whitespace here

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
Interesting, the `result of call is unused` on which the test failed isn't in the diagnosticsSema anymore. 
How did it not fail when it was first removed?

**Comment by CodaFi** [comment]
You did remove it - the test failure is because the diagnostic is now using your wording and not the old one.  You need to update that test's expectation.

**Comment by AnthonyLatsis** [comment]
I will update the expectations, but I didn't remove the `result of call is unused`, it wasn't there from the start. Probably substituted with the `result of call is unused, but produces 'T'`, which I updated.

**Comment by CodaFi** [comment]
The expectations in `-verify` mode do partial matches - which is probably why you didn't catch this the first time around.  It's very useful when you just need to make sure a diagnostic got emitted but don't necessarily care about the formatted result text.

**Comment by AnthonyLatsis** [comment]
This still means that the smoke test would fail without me opening a pull request, since the expected warning no longer existed

> The expectations in `-verify` mode do partial matches - which is probably why you didn't catch this the first time around. It's very useful when you just need to make sure a diagnostic got emitted but don't necessarily care about the formatted result text.

I invoked `lit`, does `lit` then run in `-verify` mode?

**Comment by CodaFi** [comment]
> This still means that the smoke test would fail without me opening a pull request, since the expected warning no longer existed

I don't understand what you mean.  If you ran the tests and didn't see this failure then I would check that it's even being run at all.

> I invoked lit, does lit then run in -verify mode?

Lit is just a test runner.  It interprets the `RUN` lines at the tops of files, but that's it.  The compiler itself runs in `-verify` mode (check the `RUN` line in the test you failed).

**Comment by AnthonyLatsis** [comment]
If by running the tests you mean invoking `lit` for the `test-macosx-x86_64` folder of my build directory, then yes, I did and they didn't fail.

This is the top of a failed test file:
```
// RUN: %empty-directory(%t)
// RUN: %target-swift-frontend -emit-module %S/Inputs/PrivateObjC.swift -o %t
// RUN: %target-typecheck-verify-swift -I %t -verify-ignore-unknown
```
Is `-verify-ignore-unknown` the cause is was ignored by `lit`, but not by the smoke test?

**Comment by CodaFi** [comment]
`-verify-ignore-unknown` only skips diagnostics that don't have an attached source location.  This one definitely does.

**Comment by AnthonyLatsis** [comment]
Ah, so the smoke test didn't fail before because `result of call is unused` partially matches `result of call is unused, but produces 'T'`, but not the new `result of call to function...`

But I still can't figure out why `lit` doesn't fail. I tried throwing a typo with no result. It seems it doesn't run the files that failed during the smoke test at all.

**Comment by CodaFi** [comment]
The Xcode build is not really recommended if you want to run the test suite - I personally just use it to edit files and search.  Try a ninja build.

**Comment by AnthonyLatsis** [comment]
I will. Can it coexist with the Xcode build?

Also, can you try the smoke test now that I've updated the tests?

**Comment by CodaFi** [comment]
> I will. Can it coexist with the Xcode build?

It can

@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
Hello, this thing on?

@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
The same command for a ninja build indeed does work...Looks like the best options for me is to have 2 builds, a ninja for testing and Xcode for developing.

Thanks for the advice and your patience!

**Pull request merged**